This is the initial stage of treeClassifier.

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import LabelEncoder

In [2]:
# Load data from CSV file
file_path = r'C:\Users\spthe\OneDrive\Desktop\TreeDataSet.csv'  # Update with the correct file path
data = pd.read_csv(file_path)

In [3]:
# Adjust column names
data.columns = ['TreeSpecies', 'Ratio', 'Angle']

In [4]:
# Check for data types in features and labels (for debugging)
print("Data types in features (X):")
print(data[[ 'Ratio', 'Angle']].dtypes)
print("Data type in labels (y):")
print(data['TreeSpecies'].dtype)

Data types in features (X):
Ratio    float64
Angle      int64
dtype: object
Data type in labels (y):
object


In [5]:
# Prepare features and labels
# Select features excluding column names
X = data[['Ratio', 'Angle']]

In [6]:
# Attempt to convert features to numerics (assuming numerical data)
X = X.apply(pd.to_numeric, errors='coerce').values

In [7]:
# Encode labels into integers (assuming categorical labels)
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(data['TreeSpecies'].values)

In [8]:
# Check the number of unique classes after label encoding (for debugging)
print("Number of unique classes after label encoding:", len(label_encoder.classes_))

Number of unique classes after label encoding: 3


In [9]:
# Convert labels to categorical (one-hot encoding)
encoder = OneHotEncoder(sparse=False)
y = encoder.fit_transform(y.reshape(-1, 1))  # Reshape for compatibility

In [10]:
# Check the shape of the encoded labels (for debugging)
print("Shape of encoded labels (y):", y.shape)

Shape of encoded labels (y): (90, 3)


In [11]:
# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=3)

In [12]:
# Print the shape of training features (for debugging)
print("Shape of training features (X_train):", X_train.shape)

Shape of training features (X_train): (72, 2)


In [13]:
from keras.models import Model
from keras.layers import Input, Dense

# Define the input layer with shape (2,)
input_layer = Input(shape=(2,))

# Define the layers of the model
hidden_layer1 = Dense(units=30, activation='relu')(input_layer) # First hidden layer with 30 units and ReLU activation
hidden_layer2 = Dense(units=30, activation='relu')(hidden_layer1) # Second hidden layer with 30 units and ReLU activation
output_layer = Dense(units=3, activation='softmax')(hidden_layer2) # Output layer with 5 units and softmax activation

# Define the model
custom_model = Model(inputs=input_layer, outputs=output_layer) # Creating a model with specified input and output layers

In [14]:
# Compile the model
custom_model.compile(optimizer='adam',
                     loss='categorical_crossentropy',
                     metrics=['accuracy'])

In [15]:
# Train the model with the reshaped input data
custom_model.fit(X_train, y_train, epochs=5, batch_size=30)

Epoch 1/5
3/3 [==============================] - 1s 9ms/step - loss: 12.5837 - accuracy: 0.3611
Epoch 2/5
3/3 [==============================] - 0s 5ms/step - loss: 9.4147 - accuracy: 0.3611
Epoch 3/5
3/3 [==============================] - 0s 5ms/step - loss: 6.4060 - accuracy: 0.3611
Epoch 4/5
3/3 [==============================] - 0s 4ms/step - loss: 4.6897 - accuracy: 0.3333
Epoch 5/5
3/3 [==============================] - 0s 4ms/step - loss: 4.0990 - accuracy: 0.3333


In [16]:
# Make predictions on the testing set
y_pred = custom_model.predict(X_test)

# Convert one-hot encoded predictions to class labels (optional)
y_pred_classes = np.argmax(y_pred, axis=1)

1/1 [==============================] - 0s 114ms/step


In [17]:
# Evaluate the model
test_loss, test_acc = custom_model.evaluate(X_test, y_test)
print('Test accuracy:', test_acc)

1/1 [==============================] - 0s 307ms/step - loss: 4.6506 - accuracy: 0.3333
Test accuracy: 0.3333333432674408



Now I get "0.3" as accuracy, so I change the optimizer as "rmsprop" and observe the change.

I start the procedure from beginning.

In [18]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import LabelEncoder

In [19]:
# Load data from CSV file
file_path = r'C:\Users\spthe\OneDrive\Desktop\TreeDataSet.csv'  # Update with the correct file path
data = pd.read_csv(file_path)

In [20]:
# Adjust column names
data.columns = ['TreeSpecies', 'Ratio', 'Angle']

In [21]:
# Check for data types in features and labels (for debugging)
print("Data types in features (X):")
print(data[[ 'Ratio', 'Angle']].dtypes)
print("Data type in labels (y):")
print(data['TreeSpecies'].dtype)

Data types in features (X):
Ratio    float64
Angle      int64
dtype: object
Data type in labels (y):
object


In [22]:
# Prepare features and labels
# Select features excluding column names
X = data[['Ratio', 'Angle']]

In [23]:
# Attempt to convert features to numerics (assuming numerical data)
X = X.apply(pd.to_numeric, errors='coerce').values

In [24]:
# Encode labels into integers (assuming categorical labels)
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(data['TreeSpecies'].values)

In [25]:
# Check the number of unique classes after label encoding (for debugging)
print("Number of unique classes after label encoding:", len(label_encoder.classes_))

Number of unique classes after label encoding: 3


In [26]:
# Convert labels to categorical (one-hot encoding)
encoder = OneHotEncoder(sparse=False)
y = encoder.fit_transform(y.reshape(-1, 1))  # Reshape for compatibility

In [27]:
# Check the shape of the encoded labels (for debugging)
print("Shape of encoded labels (y):", y.shape)

Shape of encoded labels (y): (90, 3)


In [28]:
# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=3)

In [29]:
# Print the shape of training features (for debugging)
print("Shape of training features (X_train):", X_train.shape)

Shape of training features (X_train): (72, 2)


In [30]:
from keras.models import Model
from keras.layers import Input, Dense

# Define the input layer with shape (2,)
input_layer = Input(shape=(2,))

# Define the layers of the model
hidden_layer1 = Dense(units=30, activation='relu')(input_layer) # First hidden layer with 30 units and ReLU activation
hidden_layer2 = Dense(units=30, activation='relu')(hidden_layer1) # Second hidden layer with 30 units and ReLU activation
output_layer = Dense(units=3, activation='softmax')(hidden_layer2) # Output layer with 5 units and softmax activation

# Define the model
custom_model = Model(inputs=input_layer, outputs=output_layer) # Creating a model with specified input and output layers

In [31]:
# Compile the model
custom_model.compile(optimizer='rmsprop',
                     loss='categorical_crossentropy',
                     metrics=['accuracy'])

In [32]:
# Train the model with the reshaped input data
custom_model.fit(X_train, y_train, epochs=5, batch_size=30)

Epoch 1/5
3/3 [==============================] - 1s 1ms/step - loss: 2.2666 - accuracy: 0.3333
Epoch 2/5
3/3 [==============================] - 0s 5ms/step - loss: 1.1795 - accuracy: 0.5000
Epoch 3/5
3/3 [==============================] - 0s 3ms/step - loss: 0.8988 - accuracy: 0.6389
Epoch 4/5
3/3 [==============================] - 0s 2ms/step - loss: 1.3032 - accuracy: 0.5278
Epoch 5/5
3/3 [==============================] - 0s 4ms/step - loss: 1.0725 - accuracy: 0.5000


In [33]:
# Make predictions on the testing set
y_pred = custom_model.predict(X_test)

# Convert one-hot encoded predictions to class labels (optional)
y_pred_classes = np.argmax(y_pred, axis=1)

1/1 [==============================] - 0s 221ms/step


In [34]:
# Evaluate the model
test_loss, test_acc = custom_model.evaluate(X_test, y_test)
print('Test accuracy:', test_acc)

1/1 [==============================] - 0s 266ms/step - loss: 0.7714 - accuracy: 0.6667
Test accuracy: 0.6666666865348816



I get the positive result model accuracy Now it was increased as "0.6".

So I keep the optimizer as "rmsprop".

Now I change the "activation='softmax'" as "activation='sigmoid'" and observe the change.

I start the procedure from beginning.

In [35]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import LabelEncoder

In [36]:
# Load data from CSV file
file_path = r'C:\Users\spthe\OneDrive\Desktop\TreeDataSet.csv'  # Update with the correct file path
data = pd.read_csv(file_path)

In [37]:
# Adjust column names
data.columns = ['TreeSpecies', 'Ratio', 'Angle']

In [38]:
# Check for data types in features and labels (for debugging)
print("Data types in features (X):")
print(data[[ 'Ratio', 'Angle']].dtypes)
print("Data type in labels (y):")
print(data['TreeSpecies'].dtype)

Data types in features (X):
Ratio    float64
Angle      int64
dtype: object
Data type in labels (y):
object


In [39]:
# Prepare features and labels
# Select features excluding column names
X = data[['Ratio', 'Angle']]

In [40]:
# Attempt to convert features to numerics (assuming numerical data)
X = X.apply(pd.to_numeric, errors='coerce').values

In [41]:
# Encode labels into integers (assuming categorical labels)
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(data['TreeSpecies'].values)

In [42]:
# Check the number of unique classes after label encoding (for debugging)
print("Number of unique classes after label encoding:", len(label_encoder.classes_))

Number of unique classes after label encoding: 3


In [43]:
# Convert labels to categorical (one-hot encoding)
encoder = OneHotEncoder(sparse=False)
y = encoder.fit_transform(y.reshape(-1, 1))  # Reshape for compatibility

In [44]:
# Check the shape of the encoded labels (for debugging)
print("Shape of encoded labels (y):", y.shape)

Shape of encoded labels (y): (90, 3)


In [45]:
# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=3)

In [46]:
# Print the shape of training features (for debugging)
print("Shape of training features (X_train):", X_train.shape)

Shape of training features (X_train): (72, 2)


In [47]:
from keras.models import Model
from keras.layers import Input, Dense

# Define the input layer with shape (2,)
input_layer = Input(shape=(2,))

# Define the layers of the model
hidden_layer1 = Dense(units=30, activation='relu')(input_layer) # First hidden layer with 30 units and ReLU activation
hidden_layer2 = Dense(units=30, activation='relu')(hidden_layer1) # Second hidden layer with 30 units and ReLU activation
output_layer = Dense(units=3, activation='sigmoid')(hidden_layer2) # Output layer with 5 units and softmax activation

# Define the model
custom_model = Model(inputs=input_layer, outputs=output_layer) # Creating a model with specified input and output layers

In [48]:
# Compile the model
custom_model.compile(optimizer='rmsprop',
                     loss='categorical_crossentropy',
                     metrics=['accuracy'])

In [49]:
# Train the model with the reshaped input data
custom_model.fit(X_train, y_train, epochs=5, batch_size=30)

Epoch 1/5
3/3 [==============================] - 1s 5ms/step - loss: 7.3796 - accuracy: 0.3333
Epoch 2/5
3/3 [==============================] - 0s 5ms/step - loss: 5.9901 - accuracy: 0.2778
Epoch 3/5
3/3 [==============================] - 0s 5ms/step - loss: 5.4281 - accuracy: 0.2500
Epoch 4/5
3/3 [==============================] - 0s 6ms/step - loss: 4.7977 - accuracy: 0.2917
Epoch 5/5
3/3 [==============================] - 0s 1ms/step - loss: 4.2366 - accuracy: 0.3056


In [50]:
# Make predictions on the testing set
y_pred = custom_model.predict(X_test)

# Convert one-hot encoded predictions to class labels (optional)
y_pred_classes = np.argmax(y_pred, axis=1)

1/1 [==============================] - 0s 113ms/step


In [51]:
# Evaluate the model
test_loss, test_acc = custom_model.evaluate(X_test, y_test)
print('Test accuracy:', test_acc)

1/1 [==============================] - 0s 222ms/step - loss: 3.3759 - accuracy: 0.3333
Test accuracy: 0.3333333432674408



Now I get the Negative result, the model accuracy was decreased as "0.3"

So I keep the "activation='softmax'"

I change the "hidden_layer1 = Dense(units=30" as "hidden_layer1 = Dense(units=60," and observe the changes.

In [52]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import LabelEncoder

In [53]:
# Load data from CSV file
file_path = r'C:\Users\spthe\OneDrive\Desktop\TreeDataSet.csv'  # Update with the correct file path
data = pd.read_csv(file_path)

In [54]:
# Adjust column names
data.columns = ['TreeSpecies', 'Ratio', 'Angle']

In [55]:
# Check for data types in features and labels (for debugging)
print("Data types in features (X):")
print(data[[ 'Ratio', 'Angle']].dtypes)
print("Data type in labels (y):")
print(data['TreeSpecies'].dtype)

Data types in features (X):
Ratio    float64
Angle      int64
dtype: object
Data type in labels (y):
object


In [56]:
# Prepare features and labels
# Select features excluding column names
X = data[['Ratio', 'Angle']]

In [57]:
# Attempt to convert features to numerics (assuming numerical data)
X = X.apply(pd.to_numeric, errors='coerce').values

In [58]:
# Encode labels into integers (assuming categorical labels)
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(data['TreeSpecies'].values)

In [59]:
# Check the number of unique classes after label encoding (for debugging)
print("Number of unique classes after label encoding:", len(label_encoder.classes_))

Number of unique classes after label encoding: 3


In [60]:
# Convert labels to categorical (one-hot encoding)
encoder = OneHotEncoder(sparse=False)
y = encoder.fit_transform(y.reshape(-1, 1))  # Reshape for compatibility

In [61]:
# Check the shape of the encoded labels (for debugging)
print("Shape of encoded labels (y):", y.shape)

Shape of encoded labels (y): (90, 3)


In [62]:
# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=3)

In [63]:
# Print the shape of training features (for debugging)
print("Shape of training features (X_train):", X_train.shape)

Shape of training features (X_train): (72, 2)


In [64]:
from keras.models import Model
from keras.layers import Input, Dense

# Define the input layer with shape (2,)
input_layer = Input(shape=(2,))

# Define the layers of the model
hidden_layer1 = Dense(units=60, activation='relu')(input_layer) # First hidden layer with 30 units and ReLU activation
hidden_layer2 = Dense(units=30, activation='relu')(hidden_layer1) # Second hidden layer with 30 units and ReLU activation
output_layer = Dense(units=3, activation='softmax')(hidden_layer2) # Output layer with 5 units and softmax activation

# Define the model
custom_model = Model(inputs=input_layer, outputs=output_layer) # Creating a model with specified input and output layers

In [65]:
# Compile the model
custom_model.compile(optimizer='rmsprop',
                     loss='categorical_crossentropy',
                     metrics=['accuracy'])

In [66]:
# Train the model with the reshaped input data
custom_model.fit(X_train, y_train, epochs=5, batch_size=30)

Epoch 1/5
3/3 [==============================] - 1s 3ms/step - loss: 2.4155 - accuracy: 0.3194
Epoch 2/5
3/3 [==============================] - 0s 173us/step - loss: 1.1970 - accuracy: 0.2917
Epoch 3/5
3/3 [==============================] - 0s 3ms/step - loss: 1.2216 - accuracy: 0.3472
Epoch 4/5
3/3 [==============================] - 0s 2ms/step - loss: 0.9477 - accuracy: 0.3889
Epoch 5/5
3/3 [==============================] - 0s 5ms/step - loss: 0.9301 - accuracy: 0.4861


In [67]:
# Make predictions on the testing set
y_pred = custom_model.predict(X_test)

# Convert one-hot encoded predictions to class labels (optional)
y_pred_classes = np.argmax(y_pred, axis=1)

1/1 [==============================] - 0s 60ms/step


In [68]:
# Evaluate the model
test_loss, test_acc = custom_model.evaluate(X_test, y_test)
print('Test accuracy:', test_acc)

1/1 [==============================] - 0s 130ms/step - loss: 0.9512 - accuracy: 0.3889
Test accuracy: 0.3888888955116272



Now I get the positive result but not as same because in the previous procedure accuracy was"0.66". 

Now the accuracy was "0.38".

I change the "hidden_layer1 = Dense(units=60" as "hidden_layer1 = Dense(units=90," and observe the changes.

In [69]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import LabelEncoder

In [70]:
# Load data from CSV file
file_path = r'C:\Users\spthe\OneDrive\Desktop\TreeDataSet.csv'  # Update with the correct file path
data = pd.read_csv(file_path)

In [71]:
# Adjust column names
data.columns = ['TreeSpecies', 'Ratio', 'Angle']

In [72]:
# Check for data types in features and labels (for debugging)
print("Data types in features (X):")
print(data[[ 'Ratio', 'Angle']].dtypes)
print("Data type in labels (y):")
print(data['TreeSpecies'].dtype)

Data types in features (X):
Ratio    float64
Angle      int64
dtype: object
Data type in labels (y):
object


In [73]:
# Prepare features and labels
# Select features excluding column names
X = data[['Ratio', 'Angle']]

In [74]:
# Attempt to convert features to numerics (assuming numerical data)
X = X.apply(pd.to_numeric, errors='coerce').values

In [75]:
# Encode labels into integers (assuming categorical labels)
label_encoder = LabelEncoder()
y = label_encoder.fit_transform(data['TreeSpecies'].values)

In [76]:
# Check the number of unique classes after label encoding (for debugging)
print("Number of unique classes after label encoding:", len(label_encoder.classes_))

Number of unique classes after label encoding: 3


In [77]:
# Convert labels to categorical (one-hot encoding)
encoder = OneHotEncoder(sparse=False)
y = encoder.fit_transform(y.reshape(-1, 1))  # Reshape for compatibility

In [78]:
# Check the shape of the encoded labels (for debugging)
print("Shape of encoded labels (y):", y.shape)

Shape of encoded labels (y): (90, 3)


In [79]:
# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=3)

In [80]:
# Print the shape of training features (for debugging)
print("Shape of training features (X_train):", X_train.shape)

Shape of training features (X_train): (72, 2)


In [81]:
from keras.models import Model
from keras.layers import Input, Dense

# Define the input layer with shape (2,)
input_layer = Input(shape=(2,))

# Define the layers of the model
hidden_layer1 = Dense(units=90, activation='relu')(input_layer) # First hidden layer with 30 units and ReLU activation
hidden_layer2 = Dense(units=30, activation='relu')(hidden_layer1) # Second hidden layer with 30 units and ReLU activation
output_layer = Dense(units=3, activation='softmax')(hidden_layer2) # Output layer with 5 units and softmax activation

# Define the model
custom_model = Model(inputs=input_layer, outputs=output_layer) # Creating a model with specified input and output layers

In [82]:
# Compile the model
custom_model.compile(optimizer='rmsprop',
                     loss='categorical_crossentropy',
                     metrics=['accuracy'])

In [83]:
# Train the model with the reshaped input data
custom_model.fit(X_train, y_train, epochs=5, batch_size=30)

Epoch 1/5
3/3 [==============================] - 0s 8ms/step - loss: 3.0841 - accuracy: 0.4028
Epoch 2/5
3/3 [==============================] - 0s 0s/step - loss: 0.9814 - accuracy: 0.3333
Epoch 3/5
3/3 [==============================] - 0s 2ms/step - loss: 0.7340 - accuracy: 0.7639
Epoch 4/5
3/3 [==============================] - 0s 5ms/step - loss: 0.9909 - accuracy: 0.6528
Epoch 5/5
3/3 [==============================] - 0s 2ms/step - loss: 0.7007 - accuracy: 0.7222


In [84]:
# Make predictions on the testing set
y_pred = custom_model.predict(X_test)

# Convert one-hot encoded predictions to class labels (optional)
y_pred_classes = np.argmax(y_pred, axis=1)

1/1 [==============================] - 0s 62ms/step


In [85]:
# Evaluate the model
test_loss, test_acc = custom_model.evaluate(X_test, y_test)
print('Test accuracy:', test_acc)

1/1 [==============================] - 0s 135ms/step - loss: 1.0926 - accuracy: 0.4444
Test accuracy: 0.4444444477558136



Now I get the positive result, the model accuracy was slightly increased as "0.44"